diff --git a/configure.ac b/configure.ac
index 7694d80d9..6d9e0e6f5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2104,10 +2104,6 @@ if test "x$GLAMOR" = xyes; then
 			[AC_DEFINE(GLAMOR_HAS_GBM_LINEAR, 1, [Have GBM_BO_USE_LINEAR])], [],
 			[#include <stdlib.h>
 			 #include <gbm.h>])
-		dnl 17.1.0 is required for gbm_bo_create_with_modifiers
-		PKG_CHECK_EXISTS(gbm >= 17.1.0,
-				 [AC_DEFINE(GBM_BO_WITH_MODIFIERS, 1, [Have gbm_bo_create_with_modifiers])],
-				 [])
 	else
 		if test "x$XORG" = xyes; then
 			AC_MSG_ERROR([Glamor for Xorg requires $LIBGBM])
